File: Scala/algorithmic_toolbox/src/main/scala-2.12/greedy/DistinctDecomposition.scala
package greedy

import scala.annotation.tailrec

object DistinctDecomposition {
  def decomposeToDistinctElems(n: Int): List[Int] = {
    @tailrec
    def loop(number: Int, smallestSummand: Int, distinctSummands: List[Int]): List[Int] = {
      if (number == 0) distinctSummands
      else {
        val nextSummand: Int = if (number > 2 * smallestSummand) smallestSummand else number
        loop(number - nextSummand, smallestSummand + 1, nextSummand :: distinctSummands)
      }
    }
    loop(n, 1, Nil)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val result: List[Int] = decomposeToDistinctElems(n)
    println(result.length)
    println(result.mkString(" "))
  }
}

File: OCaml/algorithmic_toolbox/greedy/distinct_decomposition.ml
let decompose_to_distinct_elems n =
    let rec inner number smallest_summand distinct_summands = match number with
        | 0 -> distinct_summands
        | _ -> let next_summand = if number > 2 * smallest_summand then smallest_summand else number in
               inner (number - next_summand) (smallest_summand + 1) (next_summand :: distinct_summands)
    in inner n 1 []


let () =
    let n = read_int() in
    let result = decompose_to_distinct_elems n in
    print_int (List.length result);
    print_newline();
    List.iter (fun elem -> print_int elem; print_string " ") result;
    print_newline()
